File: 백준/Silver/10799. 쇠막대기/쇠막대기.cc
#include <iostream>
#include <stack>
#include <string>
using namespace std;


int main() {
	string input = "";
	cin >> input;

	stack<char> st;
	int ans = 0;

	for (int i = 0; i < input.length(); i++) {
		if (input[i] == '(') {
			st.push('(');
		}
		else {
			if (input[i - 1] == '(') { //레이저
				st.pop();
				ans += st.size();
			}
			else { //쇠막대기 끝
				st.pop();
				ans += 1;
			}
		}
	}

	cout << ans << endl;

	return 0;
}

File: 백준/Silver/10799. 쇠막대기/쇠막대기.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String input = br.readLine();

        int result = 0;
        Stack<Character> stack = new Stack<>();

        stack.push(input.charAt(0));
        for (int i = 1; i < input.length(); i++) {
            if (input.charAt(i) == ')' && input.charAt(i - 1) == '(') {
                stack.pop();
                result += stack.size();
            } else if (input.charAt(i) == ')') {
                stack.pop();
                result++;
            } else {
                stack.push(input.charAt(i));
            }
        }
        System.out.println(result);
    }
}